validator/server_arb/machine.go

```diff
@@ -383,9 +398,16 @@ func (m *ArbitratorMachine) SetPreimageResolver(resolver GoPreimageResolver) err
 		return errors.New("machine frozen")
 	}
 	id := lastPreimageResolverId.Add(1)
-	preimageResolvers.Store(id, resolver)
+
+	refCounter := atomic.Int64{}
+	refCounter.Store(1)
+	resolverWithRefCounter := goPreimageResolverWithRefCounter{
+		resolver:   resolver,
+		refCounter: &refCounter,
+	}
+	preimageResolvers.Store(id, resolverWithRefCounter)
+
 	m.contextId = &id
```

Comment:
I just realized a potential edge case here, what if there's already a contextId set? In that case we should free the previous one. Let's add a test case for that edge case and a test case for calling Destroy twice as well if you don't already have that.

Reply:
Done.